### PR TITLE
Annotate functions with #[must_use] where appropriate

### PR DESCRIPTION
**Pull request by MalteT**
These are some initial additions for #598. I tried to be conservative, there are a lot of functions that could potentially benefit from being annotated aswell.

I haven't touched anything in `nalgebra-*` yet.

Please let me know about any concerns and improvements!

### CODE REVIEW

**Comment by sebcrozet** [comment]
Hi! Thank you for this PR. This looks all good to me :+1:

**Comment by MalteT** [comment]
Thanks! I'd like to try to achieve some kind of consistency here. I.e. annotate all `solve` functions and link to `solve_mut` if it exists. I'll try to get through the rest of the files soonish, I hope it's okay if this sits in WIP-mode for a little.

I'm also not sure how noisy this should be. Most function names are explicit enough, so that they will not benefit from `#[must_use]`. And subtle differences like `transpose` vs `transposed` make it quite clear whether one should expect mutation. I'll try to come up with a good balance but let me know if there are functions that should/shouldn't be annotated :)

**Comment by sebcrozet** [comment]
> I'd like to try to achieve some kind of consistency here. I.e. annotate all solve functions and link to solve_mut if it exists. I'll try to get through the rest of the files soonish, I hope it's okay if this sits in WIP-mode for a little.

I noticed that, this kind of consistency is very good and what you have been doing so far is great. Please take your time. I may end up forgetting about this, so don't hesitate to ping me explicitly here or on discord (I'm @sebcrozet on both) when you would like this to be reviewed or merged.

> I'm also not sure how noisy this should be. Most function names are explicit enough, so that they will not benefit from #[must_use]. And subtle differences like transpose vs transposed make it quite clear whether one should expect mutation. I'll try to come up with a good balance but let me know if there are functions that should/shouldn't be annotated :)

I feel like all the methods that take `&self` and that doesn't do any side-effect should be marked `#[must_use]`, even if their names are explicit enough. But keep in mind that it's OK for you not to address all the relevant functions in this PR. Feel free to stop whenever you want (if this starts boring you or if you no longer have the time) and let me know; then I will merge that PR and the other functions can be covered by another PR made by somebody else.

**Comment by MalteT** [comment]
Thanks for your feedback, @sebcrozet!
> I feel like all the methods that take &self and that doesn't do any side-effect should be marked #[must_use], even if their names are explicit enough. But keep in mind that it's OK for you not to address all the relevant functions in this PR.

That was my original vision, good to hear that you agree on that. I'll annotate less conservative then.

It's not the most interesting thing to do, but I'm happy to help :) I'll let you when there's an update!

File: src/linalg/cholesky.rs [hunk 1]
```diff
@@ -119,6 +119,7 @@ where
 
     /// Returns the solution of the system `self * x = b` where `self` is the decomposed matrix and
     /// `x` the unknown.
+    #[must_use = "Did you mean to use solve_mut()"]
```
Comment (vks) [inline src/linalg/cholesky.rs:122]: I think the question mark is missing.

**Comment by MalteT** [comment]
The last commits added a lot of `#[must_use]` annotations, there are probably a few false positives. I tried to keep to the following rules:
- `pub fn`s that take `&self` and return some value while having no sideeffects should be annotated
- Do not annotate `.*iter.*` functions
- Do not annotate functions with `#[doc(hidden)]`
- Do not annotate deprecated functions
- Be very aggressive

I kept the hints very basic, most functions got tagged with
```rust
#[must_use = "This function does not mutate self. You should use the return value."]
```
I'm not very happy with that hint, recommendations welcome.

Should `get_mut` functions get annotations?

These are all annotations: (Still needs some renaming)
 ```console
> rg must_use --no-filename -N --trim | sort | uniq -c | sort -nr
    448 #[must_use = "This function does not mutate self. You should use the return value."]
      8 #[must_use = "Did you mean to use inverse_mut()?"]
      7 #[must_use = "Did you mean to use two_sided_inverse_mut()?"]
      6 #[must_use = "Did you mean to use solve_mut()?"]
      6 #[must_use = "Did you mean to use conjugate_mut()?"]
      5 #[must_use = "Did you mean to use try_inverse_mut()?"]
      4 #[must_use = "Did you mean to use normalize_mut()?"]
      2 #[must_use = "This function does not mutate `val`."]
      2 #[must_use = "This function does not mutate self. Consider using the return value or dropping the function call."]
      2 #[must_use = "This function does not mutate self but returns a new clamped version."]
      2 #[must_use = "Did you mean to use try_normalize_mut()?"]
      2 #[must_use = "Did you mean to use tr_solve_lower_triangular_mut()?"]
      2 #[must_use = "Did you mean to use transpose_mut()?"]
      2 #[must_use = "Did you mean to use solve_lower_triangular_mut()?"]
      2 #[must_use = "Did you mean to use simd_try_normalize_mut()?"]
      2 #[must_use = "Did you mean to use prepend_scaling_mut()?"]
      2 #[must_use = "Did you mean to use append_scaling_mut()?"]
      2 #[must_use]
      1 #[must_use = "This function has no side effects. Consider using the return value or removing the function call."]
      1 #[must_use = "This function does not mutate the matrix. Consider using the return value or removing the function call. There's also transpose_mut() for square matrices."]
      1 #[must_use = "This function does not mutate self."]
      1 #[must_use = "strategies do nothing unless used"]
      1 #[must_use = "Did you mean to use unscale_mut()?"]
      1 #[must_use = "Did you mean to use tr_solve_upper_triangular_unchecked_mut()?"]
      1 #[must_use = "Did you mean to use tr_solve_upper_triangular_mut()?"]
      1 #[must_use = "Did you mean to use tr_solve_lower_triangular_unchecked_mut()?"]
      1 #[must_use = "Did you mean to use solve_upper_triangular_unchecked_mut()?"]
      1 #[must_use = "Did you mean to use solve_upper_triangular_mut()?"]
      1 #[must_use = "Did you mean to use solve_lower_triangular_unchecked_mut()?"]
      1 #[must_use = "Did you mean to use scale_mut()?"]
      1 #[must_use = "Did you mean to use prepend_translation_mut()?"]
      1 #[must_use = "Did you mean to use prepend_nonuniform_scaling_mut()?"]
      1 #[must_use = "Did you mean to use append_translation_mut()?"]
      1 #[must_use = "Did you mean to use append_nonuniform_scaling_mut()?"]
      1 #[must_use = "Did you mean to use ad_solve_upper_triangular_unchecked_mut()?"]
      1 #[must_use = "Did you mean to use ad_solve_upper_triangular_mut()?"]
      1 #[must_use = "Did you mean to use ad_solve_lower_triangular_unchecked_mut()?"]
      1 #[must_use = "Did you mean to use ad_solve_lower_triangular_mut()?"]
      1 #[must_use = "Did you mean to use adjoint_mut()?"]
      1 #[must_use = "Did you mean to use add_scalar_mut()?"]
 ```

[Here](https://gist.github.com/MalteT/4f6a32703f5d23a020834c4c3cd4912f) is a list of all files I checked for potential functions.

**Comment by sebcrozet** [comment]
That's great, thank you!

Maybe if we don't have any specific hint to give (wherever you added `#[must_use = "This function does not mutate self. You should use the return value."]`) we could just remove the hint altogether (`#[must_use]`) and let the compiler print its generic message.

> Should get_mut functions get annotations?

Yes, I think it would make sense for the methods like `get_mut` and `slice_mut` to be annotated as well if that's not too much trouble to add. There is no point in using them without using the return value.

**Comment by MalteT** [comment]
> Maybe if we don't have any specific hint to give (wherever you added #[must_use = "This function does not mutate self. You should use the return value."]) we could just remove the hint altogether (#[must_use]) and let the compiler print its generic message.

That's more elegant, good idea! Will do.

> Yes, I think it would make sense for the methods like get_mut and slice_mut to be annotated as well if that's not too much trouble to add. There is no point in using them without using the return value.

👍🏻 

I mentioned that I didn't annotate any function which contained `iter`, since `Iterator` is defined as
```rust
#[must_use = "iterators are lazy and do nothing unless consumed"]
pub trait Iterator {
```
and that should be enough. I hope that's enough to cause a warning!

**Comment by MalteT** [comment]
`get_mut`/`slice` fuctions should be annotated now. Checked them for mutations, found nothing.

The list of annotations is a little simpler now:
```
    466 #[must_use]
      8 #[must_use = "Did you mean to use inverse_mut()?"]
      7 #[must_use = "Did you mean to use two_sided_inverse_mut()?"]
      6 #[must_use = "Did you mean to use solve_mut()?"]
      6 #[must_use = "Did you mean to use conjugate_mut()?"]
      5 #[must_use = "Did you mean to use try_inverse_mut()?"]
      4 #[must_use = "Did you mean to use normalize_mut()?"]
      2 #[must_use = "Did you mean to use try_normalize_mut()?"]
      2 #[must_use = "Did you mean to use tr_solve_lower_triangular_mut()?"]
      2 #[must_use = "Did you mean to use transpose_mut()?"]
      2 #[must_use = "Did you mean to use solve_lower_triangular_mut()?"]
      2 #[must_use = "Did you mean to use simd_try_normalize_mut()?"]
      2 #[must_use = "Did you mean to use prepend_scaling_mut()?"]
      2 #[must_use = "Did you mean to use append_scaling_mut()?"]
      1 #[must_use = "This function does not mutate the matrix. Consider using the return value or removing the function call. There's also transpose_mut() for square matrices."]
      1 #[must_use = "This function does not mutate self."]
      1 #[must_use = "strategies do nothing unless used"]
      1 #[must_use = "Did you mean to use unscale_mut()?"]
      1 #[must_use = "Did you mean to use tr_solve_upper_triangular_unchecked_mut()?"]
      1 #[must_use = "Did you mean to use tr_solve_upper_triangular_mut()?"]
      1 #[must_use = "Did you mean to use tr_solve_lower_triangular_unchecked_mut()?"]
      1 #[must_use = "Did you mean to use solve_upper_triangular_unchecked_mut()?"]
      1 #[must_use = "Did you mean to use solve_upper_triangular_mut()?"]
      1 #[must_use = "Did you mean to use solve_lower_triangular_unchecked_mut()?"]
      1 #[must_use = "Did you mean to use scale_mut()?"]
      1 #[must_use = "Did you mean to use prepend_translation_mut()?"]
      1 #[must_use = "Did you mean to use prepend_nonuniform_scaling_mut()?"]
      1 #[must_use = "Did you mean to use append_translation_mut()?"]
      1 #[must_use = "Did you mean to use append_nonuniform_scaling_mut()?"]
      1 #[must_use = "Did you mean to use ad_solve_upper_triangular_unchecked_mut()?"]
      1 #[must_use = "Did you mean to use ad_solve_upper_triangular_mut()?"]
      1 #[must_use = "Did you mean to use ad_solve_lower_triangular_unchecked_mut()?"]
      1 #[must_use = "Did you mean to use ad_solve_lower_triangular_mut()?"]
      1 #[must_use = "Did you mean to use adjoint_mut()?"]
      1 #[must_use = "Did you mean to use add_scalar_mut()?"]
```

Feel free to request additional functions/changes. I think it should be okay like that!

**Comment by sebcrozet** [comment]
This is great, thanks a bunch @MalteT!

**Pull request merged**